plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs

```diff
@@ -255,12 +256,48 @@ which fits perfectly well into the builtins machinery.
 
 Although that becomes annoying for more complex data types. For tuples we need to provide two
 projection functions ('fst' and 'snd') instead of a single pattern matcher, which is not too
-bad, but to get pattern matching on lists we need three built-in functions: `null`, `head` and
-`tail` and to require `Bool` to be in the universe to be able to define a PLC equivalent of
+bad, but to get pattern matching on lists we need three built-in functions: @null@, @head@ and
+@tail@ and to require `Bool` to be in the universe to be able to define a PLC equivalent of
 
     matchList :: [a] -> r -> (a -> [a] -> r) -> r
     matchList xs z f = if null xs then z else f (head xs) (tail xs)
 
+If a constructor stores more than one value, the corresponding projection function packs them
+into a (possibly nested) pair, for example for
+
+    data Data
+        = Constr Integer [Data]
+        | <...>
+
+we have (pseudocode):
+
+    unConstrData (Constr i ds) = (i, ds)
+
+In order to get pattern matching over 'Data' we need a projection function per constructor as well
+as with lists, but writing (where the @Data@ suffix indicates that a function is a builtin that
+somehow corresponds to a constructor of 'Data')
+
+    if isConstrData d
+        then uncurry fConstr $ unConstrData d
+        else if isMapData d
+            then fMap $ unMapData d
+            else if isListData d
+                then fList $ unListData d
+                else <...>
+
+is tedious and inefficient and so instead we have a single @chooseData@ builtin that matches on
+its @Data@ argument and chooses the appropriate branch (type instantiations and strictness concerns
+are omitted for clarity):
+
+     chooseData
```

Comment:
This makes me wonder whether we should have `chooseList` instead of `null` for consistency (noting that `ifThenElse` is exactly `chooseBool`). That would give us a consistent story for "how to do built-in datatypes": have a `chooseFoo` function and then an `unsafeUnConstrX` function for each constructor.

Reply:
I kinda like how `if null xs then z else f (head xs) (tail xs)` reads, but I wouldn't mind using `chooseList`. But let's do that separately then.